### PR TITLE
Support specifying annotations for devworkspace pods in the DWOC

### PR DESCRIPTION
**Pull request by dkwon17**
### What does this PR do?
Provides `config.workspace.podAnnotations` field for DWOC to specify annotations for the devworkspace pods

### What issues does this PR fix or reference?
https://github.com/devfile/devworkspace-operator/issues/1270

### Is it tested? How?

1. Deploy DWO by using this catalog source and installing DWO from OperatorHub:
```
apiVersion: operators.coreos.com/v1alpha1
kind: CatalogSource
metadata:
  name: devworkspace-operator-catalog
  namespace: openshift-marketplace
spec:
  sourceType: grpc
  image: quay.io/dkwon17/devworkspace-operator-index:annotations-dwoc
  publisher: Red Hat
  displayName: DevWorkspace Operator Catalog
  updateStrategy:
    registryPoll:
      interval: 5m
```
2. Update the `devworkspace-operator-config` DWOC by adding an annotation:
```
config:
  workspace:
    podAnnotations:
      test-annotation: test
```

3. Create a workspace:
```
oc apply -f ./samples/plain.yaml -n openshift-operators 
```
4. Once the workspace pod is created, check the pod annotations and verify that `test-annoation: test` exists:
```
metadata:
  annotations:
    test-annotation: test
    ...
```
### PR Checklist

- [ ] E2E tests pass (when PR is ready, comment `/test v8-devworkspace-operator-e2e, v8-che-happy-path` to trigger)
    - [ ] `v8-devworkspace-operator-e2e`: DevWorkspace e2e test
    - [ ] `v8-che-happy-path`: Happy path for verification integration with Che

### CODE REVIEW

File: pkg/provision/workspace/deployment.go
Comment (AObuchow) [inline pkg/provision/workspace/deployment.go:200]: I noticed we are [collecting the pod additions annotations](https://github.com/devfile/devworkspace-operator/blob/main/pkg/provision/workspace/deployment.go#L302-L304) but never actually using them. Currently, this is not a big deal, since annotations are never gathered while creating the pod additions list. I believe this logic was added to `mergePodAdditions` for completeness, but hasn't been used yet.

We could merge the `podAdditions.aAnnotations` with the `workspace.Config.Workspace.PodAnnotations`, or just use the workspace.Config.Workspace.PodAnnotations alone as there won't be any actual difference. Though in the future, this might have to make this change if the podAddtions bring in any pod annotations.

File: apis/controller/v1alpha1/devworkspaceoperatorconfig_types.go
Comment (AObuchow) [inline apis/controller/v1alpha1/devworkspaceoperatorconfig_types.go:150]: We should clarify that these annotations are for DevWorkspace pods specifically, i.e.
`// PodAnnotations defines the metadata.annotations for DevWorkspace pods created by the DevWorkspace Operator.`

File: pkg/config/sync.go
Comment (AObuchow) [inline pkg/config/sync.go:387]: Changes also need to be made to [`GetCurrentConfigString()`](https://github.com/devfile/devworkspace-operator/blob/main/pkg/config/sync.go#L487) to either give the list of the config's podAnnotations, or indicate that they are being set.

**Review by AObuchow — COMMENTED** [review]
Looks great to me so far. Some minor comments. Will do some testing today, but this looks nearly ready to me :)

**Comment by AObuchow** [comment]
Tested on Minikube and works as expected :+1:

File: pkg/provision/workspace/deployment.go
Comment (dkwon17) [reply pkg/provision/workspace/deployment.go]: Sounds good, I've updated to use `mergePodAdditions`

File: pkg/config/sync.go
Comment (dkwon17) [reply pkg/config/sync.go:387]: updated

File: apis/controller/v1alpha1/devworkspaceoperatorconfig_types.go
Comment (dkwon17) [reply apis/controller/v1alpha1/devworkspaceoperatorconfig_types.go]: Updated to suggestion

**Comment by AObuchow** [comment]
@dkwon17 I haven't debugged this to verify, but I think the CI is failing because we're now creating a workspace pod that has an empty `map[string]string` of annotations, which gets converted into `nil` when the object is created on the cluster. As a result, the [SyncObjectWithCluster](https://github.com/devfile/devworkspace-operator/blob/main/pkg/provision/sync/sync.go#L67)'s diff function for pods is probably seeing that the pod on the cluster has `nil` annotations, when it expects it to be an empty `map[string]string` of annotations. This would be why we're seeing printed diffs in the CI logs:

```
    	... // 2 ignored fields
    	Spec: v1.DeploymentSpec{
    		Replicas: &1,
    		Selector: &{MatchLabels: {"controller.devfile.io/devworkspace_id": "workspace05b48af9f4d14dd4"}},
    		Template: v1.PodTemplateSpec{
    			ObjectMeta: v1.ObjectMeta{
    				... // 9 identical fields
    				DeletionGracePeriodSeconds: nil,
    				Labels:                     {"controller.devfile.io/creator": "", "controller.devfile.io/devworkspace_id": "workspace05b48af9f4d14dd4", "controller.devfile.io/devworkspace_name": "test-devworkspace"},
  - 				Annotations:                map[string]string{},
  + 				Annotations:                nil,
    				OwnerReferences:            nil,
    				Finalizers:                 nil,
    				ManagedFields:              nil,
    			},
```
I think this [function](https://github.com/devfile/devworkspace-operator/blob/main/pkg/provision/sync/diff.go#L65-L70) would be worth checking to see how it's behaving. Looking at its implementation, it doesn't seem like this would cause the `diff` to conclude that the cluster and spec objects are different, but maybe something subtle is going on.

**Comment by dkwon17** [comment]
Thanks for the feedback, I updated the PR, I learned that before, it was this check always caused the operator to try updating the deployment:
https://github.com/devfile/devworkspace-operator/blob/b49427cecf4ba036a7587be4e7319b4dfec42dcf/pkg/provision/sync/diff.go#L58

**Review by AObuchow — APPROVED** [review]
Looks good to me, awesome work @dkwon17 :grin:.
I recommend squashing some of your commits together (if possible) before we merge, but asides from that this looks good to merge :) 

We should also probably merge https://github.com/devfile/devworkspace-operator/pull/1267 before merging this PR btw.

**Comment by dkwon17** [comment]
I've squashed some commits and reviewed: https://github.com/devfile/devworkspace-operator/pull/1267

**Comment by AObuchow** [comment]
> I've squashed some commits and reviewed: #1267

Thank you :) 

Sorry, I should have specified earlier, but it's probably worth updating the commit titles (and maybe descriptions) just to clarify a bit more what these commits change.

For example, [`b7ccba1` (#1271)](https://github.com/devfile/devworkspace-operator/pull/1271/commits/b7ccba1872ca7a64fa1031a4fd15ff8a8098748f) could be titled as: "Add config.workspace.podAnnotations field to DWOC".

[`2bfdcd4` (#1271)](https://github.com/devfile/devworkspace-operator/pull/1271/commits/2bfdcd46cc86695bf48119db1d2d752292a2de5e) could be titled as "Add pod annotations from DWOC & podAdditions to devworkspace pod" and the description could include "Fix #1270" to link the original issue to the commit. 

Also.. I just had a second thought about a detail in this PR so please hold off on merging. Will write up my comment shortly.

File: controllers/workspace/devworkspace_controller.go
Comment (AObuchow) [inline controllers/workspace/devworkspace_controller.go:416]: I was on the fence about whether adding the `workspace.Config.Workspace.PodAnnotations` should be done here or somewhere closer to the devworkspace's [`getSpecDeployment()`](https://github.com/devfile/devworkspace-operator/blob/main/pkg/provision/workspace/deployment.go#L131) function, given that we have a function [`getAdditionalAnnotations()`](https://github.com/devfile/devworkspace-operator/blob/main/pkg/provision/workspace/deployment.go#L373-L389) which takes in a `DevWorkspaceWithConfig`.

However, `getAdditionalAnnotations()` is actually more like `getAdditionalDeploymentAnnotations()` since its return value is used for the [deployment annotations](https://github.com/devfile/devworkspace-operator/blob/main/pkg/provision/workspace/deployment.go#L181). You would have to rename `getAdditionalAnnotations()` -> `getAdditionalDeploymentAnnotations()`, and add append the DWOC annotations to podAdditions before [`mergePodAdditions()`](https://github.com/devfile/devworkspace-operator/blob/main/pkg/provision/workspace/deployment.go#L141) is called, the latter of which makes `getSpecDeployment()` messier.

Moreso, the area in the devworkspace controller where you're appending the annotation additions makes sense: `// Step four: Collect all workspace deployment contributions `

So I think this makes sense :+1: Just wanted to think this over outloud.

However, it _might_ be worth adding an extra commit to rename `getAdditionalAnnotations()` -> `getAdditionalDeploymentAnnotations()`. Though this is entirely optional.

Comment (dkwon17) [reply controllers/workspace/devworkspace_controller.go:416]: IMO it's ok as it is since all pod additions are collected before SyncDeploymentToCluster() in the main reconcile function https://github.com/devfile/devworkspace-operator/blob/69c21bd4ee855c9bb37db21e0eb32fcfbe36add3/controllers/workspace/devworkspace_controller.go#L471

I've added a new commit for the rename

Comment (AObuchow) [reply controllers/workspace/devworkspace_controller.go:416]: Agreed +1

**Review by AObuchow — APPROVED** [review]
LGTM :)

**Pull request merged**
